src/WebJobs.Script/Diagnostics/FunctionTraceWriterFactory.cs

```diff
@@ -22,12 +24,17 @@ public TraceWriter Create()
         {
             if (_scriptHostConfig.FileLoggingMode != FileLoggingMode.Never)
             {
-                TraceLevel functionTraceLevel = _scriptHostConfig.HostConfig.Tracing.ConsoleLevel;
-                string logFilePath = Path.Combine(_scriptHostConfig.RootLogPath, "Function", _functionName);
-                return new FileTraceWriter(logFilePath, functionTraceLevel);
+                return _writerCache.GetOrAdd(_functionName, (n) => CreateFileTraceWriter());
```

Comment:
What happens after a host restart (due to a `function.json` change or some other trigger) and execution? I suspect you'd end up returning a cached `TraceWriter` that has been disposed. 

We need a mechanism to remove the cached writer when it is disposed (which happens whenever the invoker is disposed), for two reasons:

1. So we don't returned cached disposed writers
2. So we don't hold on to writers and keep them alive for no reason (e.g. a function is disabled or deleted)